### PR TITLE
add feature: Paragraph.add_hyperlink() #74

### PR DESCRIPTION
**Pull request by tanyunshi**
I continued robertdodd's work to implement the p.add_hyperlink() method:
- to add links to paragraphs
- to add multiple texts into a link
- style the link text(run)

Documentations are also available for the dev part.

> from docx import Document
> document = Document()
> p = document.add_paragraph('some text')
> link = p.add_hyperlink('google','http://www.google.com')
google
> link.add_run('run1')
  < docx.text.run.Run object at 0x7fec7cabe908>
> link.add_run('formatted').italic = True
3

### CODE REVIEW

**Comment by scanny** [comment]
Hi @tanyunshi, I won't be able to commit this pull-request as is because it lacks tests and the API has not been vetted. But I thank you for it nonetheless and will keep it open for when we get back to hyperlinks. I'm sure it will be useful to folks in the meantime and we'll use it as a reference when we get to the implementation.

If you're interested in pursuing a commit, the best place to start is with the feature analysis. That's where the discussion that resolves the API happens, and that determines the acceptance tests which in turn of course have a strong impact on the implementation. Let me know if you're interested, happy to help you along the way with any questions on how to get into it.

**Comment by tanyunshi** [comment]
Hihi @scanny, I am interested in pursuing commits. In fact I have modified

**Comment by akobler** [comment]
Is there a plan to finish/merge this feature back to master? I am interested in inserting hyperlinks via python-docx. Btw, thanks you guys for all your work!

**Comment by scanny** [comment]
Hi all, apologies, I seem to have lost track of this one for roughly a year somehow :o: 

We're definitely interested in pursuing this feature. @tanyunshi I'll take a look at your analysis page this evening. On first glance it looks quite good. I'll provide some feedback.

Are you still interested in working on this? I know it's been some time :)

**Comment by tanyunshi** [comment]
Hi @scanny 

Glad that you are interested in this feature. ^^ 

Yes I think the hyperlink is an important feature and I will work on. Shall we start with the analysis page? :)

**Comment by scanny** [comment]
Yes, perfect. I'll take a look at it and provide comments, probably tomorrow, and we can go from there :)

File: docs/dev/analysis/features/text/hyperlink.rst [hunk 1]
```diff
@@ -0,0 +1,196 @@
+
+Hyperlink
+=========
+
+Word allows hyperlinks to be placed in the document.
+
+Hyperlink may link to a external location, for example, as an url. It may link to
+a location within the document, for example, as an bookmark. These two cases are
```
Comment (tanyunshi) [inline docs/dev/analysis/features/text/hyperlink.rst]: `a bookmark`

**Comment by tanyunshi** [comment]
Hello @scanny,
Thank you for the reviewing. :) 

I have taken your advise and improved the feature analysis. 

Just a few remarks:

As in the xml schema, a hyperlink may have multiple runs. In some use cases, we have different styles for the text in the same link(see my response above). I am not sure how much extra work it requires to implement this. However, if we make a bit more effects at the beginning, we have docx api more flexible(as it adapts the xml schema) and we have more possibilties for hyperlink stylings. 

You are right about the hyperlink access problem. I find `Paragraph.iter_run_level_items() ` a bit confusing (see [run level content](http://python-docx.readthedocs.org/en/latest/dev/analysis/features/text/run-content.html)) and I name the function as `Paragraph.iter_p_content()` in the *Candidate protocol* part. In fact I am bad at the naming thing. If you have an idea, please let me know. :)

I am not sure about the external/anchor link as it refers to a *bookmark* and so a *bookmark* should be implemented first. External links may be implemented later.

**Comment by scanny** [comment]
@tanyunshi -- Just wanted to mention to procedure for updating pull requests. To make changes, you do **not** create a new pull request (most folks seem to do this the first time :). What you do is update the branch you created the pull request with and then push that up to GitHub. The pull request is automatically updated.

That said, in this case, you'll need to create a new pull request because your pull is based on the master branch, and is also based on a different fork (robertdodd/python-docx). So you'll need to get your work into a new branch named 'feature/hyperlink' that's based on the 'feature/hyperlink' branch I've created on python-openxml. If you need help with the rebasing involved, let me know.

There are other challenges with making updates to pull requests, but we can face those when we come to them.

How familiar are you with git rebasing and with the GitHub pull process? I don't want to bore you with basics if you've already mastered them :)

**Comment by scanny** [comment]
> As in the xml schema, a hyperlink may have multiple runs. In some use cases, we have different styles for the text in the same link(see my response above). I am not sure how much extra work it requires to implement this. However, if we make a bit more effects at the beginning, we have docx api more flexible(as it adapts the xml schema) and we have more possibilties for hyperlink stylings.

I can see your use case now. It's definitely legitimate, so we'll take a look and see how much work is involved. We'll need to look at the XML Schema for the hyperlink element and then should have an idea.

> You are right about the hyperlink access problem. I find Paragraph.iter_run_level_items() a bit confusing (see run level content) and I name the function as Paragraph.iter_p_content() in the Candidate protocol part. In fact I am bad at the naming thing. If you have an idea, please let me know. :)

`.iter_run_level_items()` is the right name for that method if we implement it. There is already an `.iter_block_level_items()` method in the works on Document, and this would be analogous. The main question is what it should do with all the possible items it finds. You can see there are many possible child elements, of which run and hyperlink are just two (http://python-docx.readthedocs.org/en/latest/dev/analysis/schema/ct_p.html). The items that may appear *within* a run is something else altogether; I'm not sure what we would name that :)

For the moment I'm thinking .iter_run_level_items() would generate items of three possible types: Run, Hyperlink, and RunLevelItem. The first two would be a run and a hyperlink respectively. The last one would be used for everything else and only give access to the element (probably as .element). As other run-level items were implemented, they could be added explicitly. I'm inclined to think that Run and Hyperlink should inherit from RunLevelItem, but we can get to that later.

> I am not sure about the external/anchor link as it refers to a bookmark and so a bookmark should be implemented first. External links may be implemented later.

I don't get what you're saying here. Can you elaborate?

**Comment by tanyunshi** [comment]
> > I am not sure about the external/anchor link as it refers to a bookmark and so a bookmark should be implemented first. External links may be implemented later.

> I don't get what you're saying here. Can you elaborate?

Good evening @scanny ,

A hyperlink may be an external link, such as url, or an anchor link, such as an bookmark. In the second cas, the link does not have an r:id attribute that refers to an element in relationship. However, the link has an attribute anchor, that refers to a name of a bookmark, and the bookmark can be found somewhere in the document.

I don't think we have bookmark in python-docx.

Therefore, I remove everything about anchor links in the analyses part and think about it later. :)

**Comment by tanyunshi** [comment]
@scanny I close this PR as another one is available : https://github.com/python-openxml/python-docx/pull/278

**Pull request closed without merging**